File: comic/src/main/java/com/fryfrog/hub/comic/model/Comic.java
package com.fryfrog.hub.comic.model;

import com.fryfrog.hub.common.model.BaseEntity;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(name = "comics", indexes = {
    @Index(name = "idx_comic_title", columnList = "title"),
    @Index(name = "idx_comic_author", columnList = "author"),
    @Index(name = "idx_comic_library", columnList = "library_id")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Schema(description = "漫画（系列/作品级）")
public class Comic extends BaseEntity {

    @Schema(description = "书名", example = "进击的巨人")
    @Column(nullable = false)
    private String title;

    @Schema(description = "作者", example = "谏山创")
    private String author;

    @Schema(description = "简介")
    @Column(columnDefinition = "TEXT")
    private String overview;

    @Schema(description = "丛书名")
    private String series;

    @Schema(description = "丛书序号")
    private Integer seriesPart;

    @Schema(description = "外部刮削源 ID（Bangumi subject id），手动编辑时清除")
    private String sourceId;

    @Schema(description = "元数据来源: manual/scrape/scan", example = "scan")
    private String metadataSource;

    @Schema(description = "出版年份", example = "2009")
    private Integer pubYear;

    @Schema(description = "评分（0-5）", example = "4.5")
    private Double rating;

    @Schema(description = "作品目录或单文件路径（唯一标识）")
    @Column(unique = true, nullable = false)
    private String bookPath;

    @Schema(description = "封面本地路径")
    private String coverArtPath;

    @Schema(description = "卷/话数", example = "12")
    private Integer totalChapters;

    @Schema(description = "文件大小合计（字节）")
    private Long totalSize;

    @Schema(description = "所属资源库ID", example = "1")
    @Column(name = "library_id")
    private Long libraryId;

    @com.fasterxml.jackson.annotation.JsonGetter("coverUrl")
    public String getCoverUrl() {
        if (getId() == null || coverArtPath == null) return null;
        return com.fryfrog.hub.common.util.MediaUrlSigner.sign("/api/v1/comics/" + getId() + "/cover");
    }
}

File: fryfrog/models/comic.py
from __future__ import annotations

from sqlalchemy import Boolean, Double, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from fryfrog.db import Base
from fryfrog.models.base import TimestampMixin


class Comic(TimestampMixin, Base):
    __tablename__ = "comics"

    title: Mapped[str] = mapped_column(String, nullable=False, index=True)
    author: Mapped[str | None] = mapped_column(String, nullable=True, index=True)
    overview: Mapped[str | None] = mapped_column(Text, nullable=True)
    series: Mapped[str | None] = mapped_column(String, nullable=True)
    series_part: Mapped[int | None] = mapped_column(Integer, nullable=True)
    source_id: Mapped[str | None] = mapped_column(String, nullable=True)
    metadata_source: Mapped[str | None] = mapped_column(String, nullable=True)
    pub_year: Mapped[int | None] = mapped_column(Integer, nullable=True)
    rating: Mapped[float | None] = mapped_column(Double, nullable=True)
    book_path: Mapped[str] = mapped_column(String, unique=True, nullable=False)
    cover_art_path: Mapped[str | None] = mapped_column(String, nullable=True)
    total_chapters: Mapped[int | None] = mapped_column(Integer, nullable=True)
    total_size: Mapped[int | None] = mapped_column(Integer, nullable=True)
    library_id: Mapped[int | None] = mapped_column("library_id", Integer, nullable=True, index=True)


class ComicChapter(TimestampMixin, Base):
    __tablename__ = "comic_chapters"

    comic_id: Mapped[int] = mapped_column(ForeignKey("comics.id"), nullable=False, index=True)
    chapter_index: Mapped[int] = mapped_column("chapter_index", Integer, nullable=False)
    title: Mapped[str | None] = mapped_column(String, nullable=True)
    file_path: Mapped[str] = mapped_column(String, nullable=False)
    file_mtime: Mapped[int | None] = mapped_column(Integer, nullable=True)
    type: Mapped[str] = mapped_column(String(16), nullable=False)  # DIRECTORY/ARCHIVE
    page_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    file_size: Mapped[int | None] = mapped_column(Integer, nullable=True)


class ComicProgress(TimestampMixin, Base):
    __tablename__ = "comic_progress"
    __table_args__ = (UniqueConstraint("user_id", "comic_id", name="uq_comic_progress"),)

    user_id: Mapped[int] = mapped_column("user_id", Integer, nullable=False)
    comic_id: Mapped[int] = mapped_column(ForeignKey("comics.id"), nullable=False)
    chapter_index: Mapped[int | None] = mapped_column("chapter_index", Integer, nullable=True)
    page_index: Mapped[int | None] = mapped_column("page_index", Integer, nullable=True)
    completed: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
